File: project_euler_unified/problem_045/solution.cpp
#include <bits/stdc++.h>
using namespace std;

bool is_pentagonal(long long v) {
    if (v <= 0) return false;
    long long disc = 1 + 24 * v;
    long long s = (long long)sqrt((double)disc);
    while (s * s < disc) s++;
    while (s * s > disc) s--;
    if (s * s != disc) return false;
    return (1 + s) % 6 == 0;
}

int main() {
    // Every hexagonal number is triangular: H_m = T_{2m-1}.
    // So we just iterate hexagonal numbers and check if pentagonal.
    // Start after H_143 = 40755.
    for (long long m = 144; ; m++) {
        long long h = m * (2 * m - 1);
        if (is_pentagonal(h)) {
            cout << h << endl;
            return 0;
        }
    }
    return 0;
}

File: project_euler_unified/problem_045/solution.py
import math

def is_pentagonal(v):
    """Check if v is a pentagonal number."""
    if v <= 0:
        return False
    disc = 1 + 24 * v
    s = math.isqrt(disc)
    if s * s != disc:
        return False
    return (1 + s) % 6 == 0

def solve():
    """Find the next number after 40755 that is triangular, pentagonal,
    and hexagonal.

    Key insight: every hexagonal number H_m = m(2m-1) equals T_{2m-1},
    so all hexagonal numbers are triangular. We only need to find the
    next hexagonal number that is also pentagonal.
    """
    m = 144  # Start after H_143 = 40755
    while True:
        h = m * (2 * m - 1)
        if is_pentagonal(h):
            print(h)
            return
        m += 1

solve()
